lib/compat/wordpress-6.3/script-loader.php

```diff
@@ -83,3 +83,107 @@ function( $settings ) {
+ * Print block styles inline for non-block (classic) themes.
+ * Improves performance for classic themes by only loading the styles that are needed
+ * for the blocks that are actually used on the page, instead of loading all block styles.
+ * At the same time, this reduces layout shifts by inlining these styles
+ * instead of printing them all in the footer.
+ *
+ * NOTE: When backported to Core,
+ * this should probably be merged with the `wp_maybe_inline_styles` function.
+ *
+ * This function is hooked on `render_block`.
+ *
+ * @param string $block_content The block content.
+ * @param array  $block         The full block, including name and attributes.
+ *
+ * @return string The block content.
+ */
+function gutenberg_inline_block_styles_on_classic_themes( $block_content, $block ) {
+	// Bail early if not a classic theme.
+	if ( wp_is_block_theme() ) {
+		return $block_content;
+	}
+
+	// Get all registered blocks. Use a static var to avoid calling this
+	// for each block that gets rendered on the page.
+	static $all_registered_blocks;
+	if ( ! $all_registered_blocks ) {
+		$all_registered_blocks = WP_Block_Type_Registry::get_instance()->get_all_registered();
+	}
+
+	$style_handles = array();
+	if ( ! empty( $all_registered_blocks[ $block['blockName'] ] ) ) {
+		$style_handles = $all_registered_blocks[ $block['blockName'] ]->style_handles;
+	}
+
+	// Bail early if there are no styles for this block.
+	if ( empty( $style_handles ) ) {
+		return $block_content;
+	}
+
+	/* This filter is documented in wp-includes/script-loader.php */
+	$total_inline_limit = apply_filters( 'styles_inline_size_limit', 20000 );
+
+	// A static var to keep track of the total inlined CSS size.
+	static $total_inline_size = 0;
+
+	// Get the WP_Styles object.
+	$wp_styles = wp_styles();
+
+	foreach ( $style_handles as $style_handle ) {
+		// Skip stylesheet if it has already been printed.
+		if ( in_array( $style_handle, $wp_styles->done, true ) ) {
+			continue;
+		}
+
+		$path  = wp_styles()->get_data( $style_handle, 'path' );
+		$after = wp_styles()->get_data( $style_handle, 'after' );
+		if ( is_array( $after ) ) {
+			$after = implode( '', $after );
+		}
+
+		// Skip stylesheet if it has no path or "after" data.
+		if ( ! $path && ! $after ) {
+			continue;
+		}
+
+		// Get the CSS for this stylesheet.
+		$css = file_exists( $path ) ? file_get_contents( $path ) : '';
+
+		// Add the "after" styles.
+		$css .= $after;
+
+		// Get the size of the CSS.
+		$styles_size = strlen( $css );
+
+		// Skip if styles are empty.
+		if ( 0 === $styles_size ) {
+			$wp_styles->done[] = $style_handle;
+			continue;
+		}
+
+		// Check if adding these styles inline will exceed the limit.
+		// If it does, then bail early.
+		if ( $total_inline_size + $styles_size > $total_inline_limit ) {
+			continue;
+		}
+
+		$wp_styles->done[] = $style_handle;
+		$block_content    .= '<style id="' . $style_handle . '-inline-css">' . $css . '</style>';
+		wp_dequeue_style( $style_handle );
```

Comment:
I have a concern about this in that for some sites it could cause the stylesheets to not end up on the frontend at all. For example, in my [syntax-highlighting-code-block](https://github.com/westonruter/syntax-highlighting-code-block) plugin, it currently is waiting to print the stylesheets for syntax highlighting until a Code block appears on the page. The `link[rel=stylesheet]` tag and inline `style` tag are injected right after the `<pre>` start tag for the first Code block. However, this has caused headaches for some sites, where the stylesheets aren't getting printed at all: https://github.com/westonruter/syntax-highlighting-code-block/issues/286

The reason is that some plugins, like the Fixed TOC table of contents plugin, will call `do_blocks()` on the content beforehand (e.g. by [applying `the_content` filters](https://github.com/openlab-at-city-tech/openlab/blob/a77f31dacc628d5680d791340ee5c9cc421d5a33/wp-content/plugins/fixed-toc/frontend/class-frontend-control.php#LL139C14-L139C54)) to obtain content for obtaining the table of contents. There are a [several thousand plugins](https://wpdirectory.net/search/01GZ4WVY15DDXRY7YPT6FHRG0H) that manually apply `the_content` filters, and [hundreds of others](https://wpdirectory.net/search/01GZ4X14XY3JBW76S55NAVPABT) that manually call `do_blocks()`. I'm concerned that they will cause this stylesheet processing to break.

Reply:
Moved to https://github.com/WordPress/gutenberg/issues/49927#issuecomment-1528174046